userspace/otpilot/src/main.rs

```diff
@@ -225,29 +226,113 @@ impl<'a> SpiProcessor<'a> {
         }
     }
 
-    fn process_spi_write(&mut self, addr: u32, data: &[u8]) -> Result<(), FromWireError> {
-        match addr {
-            0x02000000 => {
-                self.process_spi_payload(data)
+    fn spi_host_write_enable(&mut self) -> Result<(), FromWireError> {
+        let header = flash::Header::<u32> {
+            opcode: OpCode::WriteEnable,
+            address: None,
+        };
+        let data : [u8; 0] = [0; 0];
+        self.spi_host_send(&header, &data)
+    }
+
+    fn spi_host_send<H>(&mut self, header: &H, data: &[u8]) -> Result<(), FromWireError>
+    where H: flash::SpiHeader + ToWire {
+        let mut console = Console::new();
+        let mut tx_buf = [0xff; spi_host::MAX_READ_BUFFER_LENGTH];
+        let tx_len : usize;
+        {
+            let mut tx_cursor = SpiutilsCursor::new(&mut tx_buf);
+            if let Err(why) = header.to_wire(&mut tx_cursor) {
+                writeln!(console, "Host: Could not store header: {:?}", why);
+                return Err(FromWireError::OutOfRange);
             }
-            _ => {
-                Err(FromWireError::OutOfRange)
+            if header.get_opcode().has_dummy_byte() {
+                // Skip dummy byte
+                if let Err(why) = tx_cursor.write_bytes(&[1; 0]) {
+                    writeln!(console, "Host: Could not store dummy byte: {:?}", why);
+                    return Err(FromWireError::OutOfRange);
+                }
+            }
+
+            if let Err(why) = tx_cursor.write_bytes(&data) {
+                writeln!(console, "Host: Could not store data: {:?}", why);
+                return Err(FromWireError::OutOfRange);
             }
+
+            tx_len = tx_cursor.consumed_len()
         }
```

Comment:
Why are lines 243-263 in an indented block?

Reply:
We need to lose `tx_cursor` since it holds a `&mut` on `tx_buf`. We need to get a new `&mut` on `tx_buf` in line 268.

Reply:
I moved the code into its own function, I think this is cleaner now.